File: apps/mobile/src/types/order.ts
import type { IsoDate, Money, UUID } from './common';

export type OrderStatus =
  | 'CREATED'
  | 'WAITING_PAYMENT'
  | 'CONFIRMED'
  | 'IN_PREPARATION'
  | 'SHIPPED'
  | 'DELIVERED'
  | 'CANCELLED'
  | 'REFUNDED';

export type PaymentStatus =
  | 'SIMULATED'
  | 'PENDING'
  | 'APPROVED'
  | 'REJECTED'
  | 'REFUNDED'
  | 'CANCELLED';

export interface OrderItem {
  id: UUID;
  listingId: UUID;
  sellerId: UUID;
  bookTitle: string;
  bookCoverUrl?: string;
  quantity: number;
  unitPrice: Money;
  subtotal: Money;
}

export interface Order {
  id: UUID;
  code: string;
  buyerId: UUID;
  items: OrderItem[];
  status: OrderStatus;
  paymentStatus: PaymentStatus;
  subtotalAmount: Money;
  shippingAmount: Money;
  totalAmount: Money;
  createdAt: IsoDate;
}

File: apps/api/src/main/kotlin/com/culturaz/api/orders/Order.kt
package com.culturaz.api.orders

import com.culturaz.api.users.Address
import com.culturaz.api.users.User
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.ForeignKey
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.math.BigDecimal
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "orders")
class Order(
    @Id
    @Column(name = "id", nullable = false, updatable = false)
    var id: UUID = UUID.randomUUID(),

    @Column(name = "code", nullable = false, length = 20, unique = true)
    var code: String,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "buyer_id", nullable = false, foreignKey = ForeignKey(name = "fk_orders_buyer"))
    var buyer: User,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    var status: OrderStatus = OrderStatus.CREATED,

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_status", nullable = false, length = 30)
    var paymentStatus: PaymentStatus = PaymentStatus.PENDING,

    @Column(name = "subtotal_amount", nullable = false, precision = 12, scale = 2)
    var subtotalAmount: BigDecimal,

    @Column(name = "shipping_amount", nullable = false, precision = 12, scale = 2)
    var shippingAmount: BigDecimal = BigDecimal.ZERO,

    @Column(name = "total_amount", nullable = false, precision = 12, scale = 2)
    var totalAmount: BigDecimal,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shipping_address_id", foreignKey = ForeignKey(name = "fk_orders_address"))
    var shippingAddress: Address? = null,

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    var items: MutableList<OrderItem> = mutableListOf(),

    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now(),
)
